src/dp_nat.c

```diff
@@ -721,6 +721,7 @@ int dp_remove_network_snat_port(const struct flow_value *cntrack)
 	struct dp_port *created_port;
 	int ret;
 
+	// what if cntrack->flow_key[DP_FLOW_DIR_REPLY].l3_type == RTE_ETHER_TYPE_IPV6?
 	portoverload_tbl_key.nat_ip = cntrack->flow_key[DP_FLOW_DIR_REPLY].l3_dst.ip4;
 	portoverload_tbl_key.dst_ip = cntrack->flow_key[DP_FLOW_DIR_ORG].l3_dst.ip4;
 	portoverload_tbl_key.nat_port = cntrack->flow_key[DP_FLOW_DIR_REPLY].port_dst;
```

Comment:
This one, an all following ones in `dnat_node` and `snat_node` are NAT related and I think they simply work because NAT only uses IPv4?

But unless you are familiar with the code base this will never be clear and will always seem problematic, so maybe just return error when the type is wrong? In this file it's no problem then the other ones are a hot path...

Reply:
Exactly. NAT only uses IPv4.
IPv6 is NATless apart from the fact that NAT64 but it works totally different then the ipv6 NAT.

Returning error for IPv6 should be fine. You can still check the call chain to have a rough idea what might be happening if you return error.

Reply:
There is a problem in NAT64 implementation here. in the NAT64 test `test_vf_to_pf.py::test_vf_to_pf_network_nat_icmpv6` (you can only run this one to see), the flow is not deleted here, because `portoverload_tbl_key.dst_ip = cntrack->flow_key[DP_FLOW_DIR_ORG].l3_dst.ip4;` is not actually IPv4, it's the NAT64 address.

If I copy the code from `dp_allocate_network_snat_port()` where you pull the IPv4 out from NAT64, the address is converted right, but unfortnuately there must be som followup problem because `portoverload_tbl_key.dst_port = flow_key_org->port_dst;` is ZERO for some reason, thus the flow is never deleted.

There is no error or check for it in the test suite so tests behave normally, but you can easily add debug outputs to see that a NAT64 flow is added in `dp_allocate_network_snat_port()` but never deleted in `dp_remove_network_snat_port` because of this.

Reply:
I need to revise/correct part of my previous statement. That is what happens, if you look at the code part, after a long while.
NAT is actually used by ipv6 in a similar way as in ipv4 for the NAT ipv4 address used in NAT64. Probably you figured that out, already, because of the code manipulating source ports.
To be honest, I can not tell you at the moment how to deal with the bug here but if it hinders your refactoring, can you leave this part as it is and we can fix it in a follow-up ? I will try to give it a more thought tomorrow.

Reply:
It's not problem for the refactoring (tests still work), I can then create a separate issue on the ne code.

Reply:
Ok. This seems to be broken already in ipv4 and ipv6 introduction doesnt seem to make it "worse". (ipv6 seems to be putting the wrong dst ip address in the key in addition to what was already broken)
The same problem can be observed with the ipv4 test `test_vf_to_pf_network_nat_icmp`
I will open a bug to track it and I think @byteocean can shed a light on this.